curator/src/https.py

```diff
@@ -0,0 +1,59 @@
+import urllib2
+import httplib
+import ssl
+import socket
+
+DEFAULT_HTTP_TIMEOUT = 10
+
+class HTTPSClientAuthHandler(urllib2.HTTPSHandler):
+    '''
+    Allows sending a client certificate with the HTTPS connection.
+    This version also validates the peer (server) certificate
+    '''
+    def __init__(self, key=None, cert=None, ca_certs=None, ssl_version=None, ciphers=None):
+        urllib2.HTTPSHandler.__init__(self)
+        self.key = key
+        self.cert = cert
+        self.ca_certs = ca_certs
+        self.ssl_version = ssl_version
+        self.ciphers = ciphers
+
+    def https_open(self, req):
+        return self.do_open(self.get_connection, req)
+
+    def get_connection(self, host, timeout=DEFAULT_HTTP_TIMEOUT):
+        return HTTPSConnection( host,
+                key_file = self.key,
+                cert_file = self.cert,
+                timeout = timeout,
+                ciphers = self.ciphers,
+                ca_certs = self.ca_certs,
+                ssl_version = self.ssl_version )
+
+class HTTPSConnection(httplib.HTTPSConnection):
+    '''
+    Overridden to allow peer certificate validation, configuration
+    of SSL/ TLS version and cipher selection.  See:
+    http://hg.python.org/cpython/file/c1c45755397b/Lib/httplib.py#l1144
+    and `ssl.wrap_socket()`
+    '''
+    def __init__(self, host, **kwargs):
+        self.ciphers = kwargs.pop('ciphers', None)
+        self.ca_certs = kwargs.pop('ca_certs', None)
+        self.ssl_version = kwargs.pop('ssl_version', ssl.PROTOCOL_SSLv23)
+
+        httplib.HTTPSConnection.__init__(self,host,**kwargs)
+
+    def connect(self):
+        sock = socket.create_connection( (self.host, self.port), self.timeout )
+
+        if self._tunnel_host:
+            self.sock = sock
+            self._tunnel()
+
+        self.sock = ssl.wrap_socket( sock,
```

Comment:
Don't use ``ssl.wrap_socket``. It doesn't validate host names and is therefore vulnerable to MitM attacks. You want to use a SSLContext object from ``ssl.create_default_context`` and then use its wrap_socket() method with ``server_hostname``.